docs/rfc/rfc-018-bls-agg-exploration.md

```diff
@@ -0,0 +1,400 @@
+and verifying that a given solution is correct.
+
+Pairing-based cryptography works by operating on mathematical functions that
+satisfy the property of 'bilinear mapping'. This property is satisfied for
+functions `e` with values `P`, `Q`, and `R` where `e(P, Q + R) = e(P, Q) * e(P, R)`
+and `e(P + Q, R) = e(P, R) * e(Q, R)`. The most familiar example of this is
+exponentiation. Written in common notation, P^(Q+R) = P^Q * P^R.
+
+Pairing-based elliptic-curve cryptography creates a bilinear mapping using
+an elliptic-curve. With some original curve, you can define two groups,
+`G1` and `G2` which are points of the original curve _modulo_ different values.
+Finally, you define a third group `Gt`, where points from `G1` and `G2` satisfy
+the property of bilinearity with `Gt`. In this scheme, the function `e` takes
+as inputs points in `G1` and `G2` and outputs values in `Gt`. Succintly, given
+some point `P` in `G1` and some point `Q` in `G1`, `e(P, Q) = C` where `C` is in `Gt`.
+You can efficiently compute the mapping of points in `G1` and `G2` into `Gt`,
+but you cannot efficiently determine what points were summed and paired to
+produce the value in `Gt`.
+
+Functions are then defined to map digital signatures, messages, and keys into
+and out of points of `G1` or `G2` and signature verification is the process
+of calculating if a set of values representing a message, public key, and digital
+signature produce the same value in `Gt` through `e`.
+
+Signatures can be created as either points in `G1` with public keys being
+created as points in `G2` or vice versa. For the case of BLS12-381, the popular
+curve used, points in `G1` are represented with 48 bytes and points in `G2` are
+represented with 96 bytes. It is up to the implementer of the cryptosystem to
+decide which should be larger, the public keys or the signatures.
+
+BLS signatures rely on pairing-based elliptic-curve cryptography to produce
+various types of signatures. For a more in-depth but still high level discussion
+pairing-based elliptic-curve cryptography, see Vitalik Buterin's post on
+[Exploring Elliptic Curve Pairings][vitalik-curve-post]. For much more in
+depth discussion, see the specific paper on BLS12-381, 
+[Compact Multi-Signatures for Smaller Blockchains][multi-signatures-smaller-blockchains].
+
+### Adoption
+
+BLS signatures have already gained traction within several popular projects.
+
+* Algorand is working on an implementation.
+* [Zcash][zcash-adoption] has adopted BLS12-381 into the protocol.
+* [Ethereum 2.0][eth-2-adoption] has adopted BLS12-381 into the protocol.
+* [Chia Network][chia-adoption] has adopted BLS for signing blocks.
+* [Ostracon][line-ostracon-pr], a fork of Tendermint has adopted BLS for signing blocks.
+
+### What systems may be affected by adding aggregated signatures?
+
+#### Gossip
+
+Gossip could be updated to aggregate vote signatures during a consensus round.
```

Comment:
Apart from the overhead, this would also be a significant expansion of the gossip protocol's sphere of responsibility: In addition to knowing who the peers are, and what messages to send and/or request from them, it would also have to gain a lot more knowledge about message contents, and integrate with code for constructing and aggregating cryptographic signatures. That further connects it to concerns about key management that seem out-of-place for a communications layer.

Reply:
Yes, I definitely agree that it would expand the responsibilities of the gossip layer. _If_ that expansion and complexity confers serious benefit we should consider it.

Reply:
Well, this is only partially true. The consensus layer only relays messages that were verified by the consensus layer, that is, which signatures are valid. While this verification is done gone at "gossip level", it determines the operation of the gossip layer. In fact, in the current implementation, the separation between gossip and protocol layers is merely in terms of organization, as the gossip layer does not operate independently of the associate protocol.

Reply:
Sure, I agree that the gossip layer holds some responsibility for _what_ to relay, it does not, at the moment, modify the data it receives from consensus. It simply determines _who_ to send _what_ to. This would include additional capabilities and responsibilities for aggregating and storing data that it currently does not perform.